pkg/subscribe/debouncer.go

```diff
@@ -0,0 +1,87 @@
+package subscribe
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/rancher/apiserver/pkg/types"
+)
+
+type DebouncerState int
+
+const (
+	// The first notification is always sent right away, no need to wait
+	FirstNotification DebouncerState = iota
+	TimerStarted
+	TimerStopped
+)
+
+type debouncer struct {
+	lock sync.Mutex
+
+	timer        *time.Timer
+	debounceRate time.Duration
+
+	inCh  chan types.APIEvent
+	outCh chan types.APIEvent
+}
+
+func newDebouncer(debouceRate time.Duration, eventsCh chan types.APIEvent) *debouncer {
+	d := &debouncer{
+		debounceRate: debouceRate,
+		timer:        time.NewTimer(debouceRate),
+		inCh:         eventsCh,
+		outCh:        make(chan types.APIEvent),
+	}
+	d.timer.Stop()
+	return d
+}
+
+func (d *debouncer) Run(ctx context.Context) {
+	state := FirstNotification
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			break loop
+		case ev, ok := <-d.inCh:
+			if ev.Error != nil {
+				ev.Name = string(SubscriptionModeNotification)
+				d.outCh <- ev
+				break loop
+			}
+
+			if !ok {
+				break loop
+			}
+
+			d.lock.Lock()
+			switch state {
+			case FirstNotification:
+				d.outCh <- types.APIEvent{
+					Name: string(SubscriptionModeNotification),
+				}
```

Comment:
This could be a lack of understanding on my part, but looks like we are dropping a lot of information here right? Should we adopt the same approach as above and just set the name?

```go
ev.Name = string(SubscriptionModeNotification)
d.outCh <- ev
```

Reply:
(Look at the example in the PR description above)

Reply:
Ok cool works for me :+1: